navigation/navigation-runtime/src/wasmJsMain/kotlin/androidx/navigation/BrowserHistory.wasmJs.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.navigation
+
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.launch
+import org.w3c.dom.PopStateEvent
+import org.w3c.dom.Window
+import org.w3c.dom.events.Event
+
+suspend fun Window.bindToNavigation(navController: NavController) {
+    coroutineScope {
```

Comment:
We can try to share the implementation code by adding a few internal external interfaces in webMain:

```kotlin
internal external interface WindowW3C {
    val location: LocationW3C
    val history: HistoryW3C
    ...
}

internal external iterface LocationW3C {
   val origin: String
   val pathname: String
}

internal external interface HistoryW3C {
...
}
```

The benefit would be is that we won't have to change the code in 2 places.

The function still won't be common and will use the original `Window` type which can be casted to our new internal type.

Reply:
good idea. I've implemented